d2/src/main/java/com/linkedin/d2/xds/XdsToD2PropertiesAdaptor.java

```diff
@@ -0,0 +1,143 @@
+        Executors.newSingleThreadScheduledExecutor());
+
+    XdsToD2PropertiesAdaptor adaptor = new XdsToD2PropertiesAdaptor(xdsClient);
+    adaptor.listenToService("d2-service-polarisfoo1.prod.linkedin.com");
+
+    while (true)
+    {
+    }
+  }
+
+  public void setUriEventBus(PropertyEventBus<UriProperties> uriEventBus)
+  {
+    _uriEventBus = uriEventBus;
+  }
+
+  public void setServiceEventBus(PropertyEventBus<ServiceProperties> serviceEventBus)
+  {
+    _serviceEventBus = serviceEventBus;
+  }
+
+  public void setClusterEventBus(PropertyEventBus<ClusterProperties> clusterEventBus)
+  {
+    _clusterEventBus = clusterEventBus;
+  }
+
+  public void listenToCluster(String clusterName)
+  {
+    _xdsClient.watchCdsResource(clusterName, cdsUpdate ->
+    {
+      Map<String, Value> clusterPropertiesMap = cdsUpdate.getClusterProperties();
+      if (_clusterEventBus != null)
+      {
+        _clusterEventBus.publishInitialize(clusterName,
+            XdsToClusterPropertiesAdaptor.toClusterProperties(clusterPropertiesMap));
+      }
+
+      String serviceName = cdsUpdate.getEdsServiceName();
+      _xdsClient.watchEdsResource(serviceName, edsUpdate ->
+      {
+        List<LocalityLbEndpoints> endpoints = edsUpdate.getLocalityLbEndpoints();
+
+        if (_uriEventBus != null)
+        {
+          _uriEventBus.publishInitialize(clusterName, XdsToUriPropertiesAdaptor.toUriProperties(endpoints));
+        }
+      });
+    });
+  }
+
+  public void listenToService(String serviceName)
+  {
+    _xdsClient.watchLdsResource(serviceName + "_18080", ldsUpdate ->
+    {
+      HttpConnectionManager httpConnectionManager = ldsUpdate.getHttpConnectionManager();
+      String routeConfigName = httpConnectionManager.getRds().getRouteConfigName();
+
+      _xdsClient.watchRdsResource(routeConfigName, rdsUpdate ->
+      {
+        Map<String, Value> servicePropertiesMap = rdsUpdate.getServiceProperties();
+        if (_serviceEventBus != null)
+        {
+          _serviceEventBus.publishInitialize(serviceName,
+              XdsToServicePropertiesAdaptor.toServiceProperties(servicePropertiesMap));
+        }
+
+        List<String> clusterNames = new ArrayList<>();
+        for (VirtualHost virtualHost : rdsUpdate.getVirtualHosts())
+        {
+          for (Route route : virtualHost.getRoutesList())
+          {
+            String clusterName = route.getRoute().getCluster();
+            clusterNames.add(clusterName);
+          }
+        }
+
+        for (String clusterName : clusterNames)
+        {
+          listenToCluster(clusterName);
+        }
+      });
+    });
+  }
+
+  private static final class XdsToUriPropertiesAdaptor
+  {
+    static UriProperties toUriProperties(List<LocalityLbEndpoints> endpoints)
+    {
+      // TODO: implement this
+      return null;
+    }
+  }
+
+  private static final class XdsToServicePropertiesAdaptor
+  {
+    static ServiceProperties toServiceProperties(Map<String, Value> serviceProperties)
```

Comment:
just a reminder that we need to convert to ServiceStoreProperties and ClusterStoreProperties to include canary and failout configs.

Reply:
Sounds good.